用自己的数据，复现参考文献的流程

参考文献：[***2025 - A metagenome-wide study of the gut virome in chronic kidney disease***](https://pmc.ncbi.nlm.nih.gov/articles/PMC11780533/)

#### **Identification and processing of viral sequences**

The workflow of viral identification is shown in [Figure S1](https://pmc.ncbi.nlm.nih.gov/articles/PMC11780533/#SM0)A. Metagenomically assembled contigs were recognized as viral sequences based on their sequence features and homology to known viral genomes. Raw viral contigs were identified when they satisfied one of the following criteria: 1) identified as a virus in VIBRANT v1.2.1 [43](https://pmc.ncbi.nlm.nih.gov/articles/PMC11780533/#B43) with default parameters (-meta mode); 2) containing a greater number of viral genes than microbial genes based on searches against the CheckV marker gene set [44](https://pmc.ncbi.nlm.nih.gov/articles/PMC11780533/#B44); or 3) achieving a score >0.9 and *P* <0.01 in DeepVirFinder v1.0 [45](https://pmc.ncbi.nlm.nih.gov/articles/PMC11780533/#B45). Contigs that were recognized as “undetermined” sequences by CheckV were discarded. Parallelly, raw provirus sequences were extracted from the contigs by CheckV, with those shorter than 5kb removed. These procedures generated a total of 178,097 candidate viral sequences (132,172 viral contigs and 45,925 proviruses). To decontaminate the viral sequences, according to the previous study [22](https://pmc.ncbi.nlm.nih.gov/articles/PMC11780533/#B22), [46](https://pmc.ncbi.nlm.nih.gov/articles/PMC11780533/#B46), we searched the bacterial universal single-copy orthologs (BUSCO) [47](https://pmc.ncbi.nlm.nih.gov/articles/PMC11780533/#B47) within the raw viral sequence using hmmsearch [48](https://pmc.ncbi.nlm.nih.gov/articles/PMC11780533/#B48) with default options and calculated the BUSCO ratio as the number of BUSCOs relevant to the total number of genes in each viral sequence. High-contaminated viral sequences with a BUSCO ratio ≥5% were removed, resulting in 170,759 sequences considered as the final viral sequences from the metagenomic samples.

The viral sequences were de-replicated based on the following steps: 1) all viral sequences were pairwise aligned using BLASTn v2.9.0 with the options “-evalue 1e-10 -word_size 20 -num_alignments 99999”; 2) viral sequences which shared 95% nucleotide identity across 75% of their length were clustered into a viral operational taxonomic unit (vOTU) using custom scripts(https://github.com/RChGO/virusDectect); 3) The longest viral sequence was considered as the representative sequence for each vOTU. Additionally, shared vOTUs between different gut virus collections (i.e., Gut Virome Database [22](https://pmc.ncbi.nlm.nih.gov/articles/PMC11780533/#B22) and Gut Phage Database [23](https://pmc.ncbi.nlm.nih.gov/articles/PMC11780533/#B23)) were identified following the same steps as above, and the combined nonredundant vOTU catalogue (n = 45,849) was generated accordingly.

#### **Viral taxonomic classification and functional annotation**

Viral proteins of vOTUs were predicted using Prodigal v2.6.3 49. We compiled a reference database by aggregating protein sequences from three viral databases: the Virus-Host DB 50 (downloaded in May 2021), crAss-like phage proteins from Guerin's study 51, and the protein catalogue from Benler's study 52. For accurate family-level taxonomic classification of viruses, we aligned the proteins of all known viral sequences from the NCBI-RefSeq database against the combined reference database using DIAMOND 53 with the options “--query-cover 50 --subject-cover 50 --id 30 --min-score 50 --max-target-seqs 10”. A viral sequence was annotated to a viral family-level taxon when over 25% of its proteins matched that family. This approach obtained an accuracy of 98.6% for family-level classification of the viruses from the NCBI-RefSeq database, which we applied to the taxonomic classification of the vOTUs. For functional analysis of viral populations, we performed BLAST searches of protein sequences of all vOTUs against the KEGG (Kyoto Encyclopedia of Genes and Genomes) database (downloaded in December 2020) using DIAMOND with the following options “--query-cover 50 --subject-cover 50 -e 1e-5 --min-score 50 --max-target-seqs 50”. The matched protein was annotated to a KEGG orthologue (KO) based on the best-hit protein.

##### 流程步骤

- **cluster**：1) all viral sequences were pairwise aligned using BLASTn v2.9.0 with the options “-evalue 1e-10 -word_size 20 -num_alignments 99999”; 2) viral sequences which shared 95% nucleotide identity across 75% of their length were clustered into a viral operational taxonomic unit (vOTU) using custom scripts (https://github.com/RChGO/virusDectect); 3) The longest viral sequence was considered as the representative sequence for each vOTU.
- **Prodigal**：（v2.6.3）Viral proteins of vOTUs were predicted using Prodigal
- **three viral databases**：the [Virus-Host DB](https://www.mdpi.com/1999-4915/8/3/66) ([downloaded in May 2021](https://www.genome.jp/ftp/db/virushostdb/old/release205/)), *crAss-like* phage proteins from [Guerin's study](https://www.cell.com/cell-host-microbe/fulltext/S1931-3128(18)30524-9)（[或者](https://www.nature.com/articles/s41467-021-21350-w)）, and the protein catalogue from [Benler's study](https://www.cell.com/cell-host-microbe/fulltext/S1931-3128(18)30524-9?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS1931312818305249%3Fshowall%3Dtrue)
- **DIAMOND**：For accurate family-level taxonomic classification of viruses, we aligned the proteins of all known viral sequences from the NCBI-RefSeq database against the combined reference database using DIAMOND with the options “`--query-cover 50 --subject-cover 50 --id 30 --min-score 50 --max-target-seqs 10`”. A viral sequence was annotated to a viral family-level taxon when over 25% of its proteins matched that family. This approach obtained an accuracy of 98.6% for family-level classification of the viruses from the NCBI-RefSeq database, which we applied to the taxonomic classification of the vOTUs.

#### 运行此流程

`snakemake -s pipeline.smk --cores 5 --use-conda`

#### 输出示例

| votu_id                   | family         | lineage                                                      |
| ------------------------- | -------------- | ------------------------------------------------------------ |
| F1_1A_contig_996\|\|full  | Unclassified   | Duplodnaviria; Heunggongvirae; Uroviricota; Caudoviricetes; Caudovirales; ; |
| F1_1A_contig_5049\|\|full | Herelleviridae | Viruses; Duplodnaviria; Heunggongvirae; Uroviricota; Caudoviricetes; Herelleviridae; Bastillevirinae; Caeruleovirus; unclassified Caeruleovirus |