---
title: "03_community_types"
output: html_document
---
```{r}
# 安装 R 包
micromamba create -n r r-base=4.3.0 bioconductor-microbiomemarker bioconductor-deseq2 
micromamba activate r

micromamba install r-devtools r-biotyper r-reshape r-writexls r-dendsort r-seriation \
                   r-taxonomizr r-esquisse r-tidyverse r-rstatix r-PMCMR r-clValid \
                   r-philentropy bioconductor-microbiome r-RcppArmadillo=0.9.900.3.0 -y
```

```{r}
library(ggplot2)
library(dplyr)
library(plyr)
library(reshape)
library(reshape2)
library(scales)
library(viridis)
library(readr)
library(WriteXLS)
library(readxl)
library(remotes)
library(devtools)
library(BiocManager)
library(ComplexHeatmap)
library(phyloseq)
library(ALDEx2)
library(RSQLite)
library(RcppArmadillo)
library(DESeq2)
library(microbiomeMarker)
library(vegan)
library(dendsort)
library(ape)
library(seriation)
library(taxonomizr)
library(esquisse)
library(tidyverse)
library(grid)
library(coin)
library(rstatix)
library(PMCMR)
library(BiotypeR)
library(DirichletMultinomial)
library(shiny)
library(fpc)
library(clValid)
library(cluster) 
library(philentropy)
```

```{r}
# 检查当前工作目录
getwd()
# 如果需要，可以改变工作目录
setwd("/home/shijiabin/2025_2ME/11_viral_community_types")
getwd()
```

```{r}
#################################### 1. Create a phylseq objects ####################################
####################################
# 1.1 Genus-like groups input
# vOTU 绝对丰度表 (已稀释,消除测序深度的影响)
votu_abundance_file <- "01_raw_counts/votu_raw_counts.tsv"
abuntable <- read.table(votu_abundance_file, sep = "\t", header = TRUE, stringsAsFactors = FALSE)

# 读取 aai 聚类结果文件
genus_clusters_file <- "02_genus_clusters/genus_clusters.txt"
lines <- readLines(genus_clusters_file)

# 构建 vOTU -> group 映射
votu_to_group <- c()
for (i in seq_along(lines)) {
  votus <- strsplit(lines[i], "\t")[[1]]
  votus <- votus[votus != ""]
  votu_to_group[votus] <- paste0("Genus",i)
}

# 添加 group 列
abuntable$Genus_like_groups <- votu_to_group[abuntable$vOTU]

####################################
# 1.2 汇总分类注释
# 分类注释表
taxtable_file <- "/home/shijiabin/2025_2ME/05_taxonomic_classification/votus_taxonomy.tsv"
taxtable <- read.table(taxtable_file, sep = "\t", header = TRUE, stringsAsFactors = FALSE)

# 添加 Genus_like_groups 列, 绝对丰度列
Mastertable <- merge(taxtable, abuntable, by=1, all=F)

# votu 列做行名 (方便后续的表格操作, phyloseq能够识别)
rownames(Mastertable) <- Mastertable$vOTU

# 根据原代码 https://github.com/Matthijnssenslab/IBDVirome/blob/main/IBDBiologicals/Code/PartVIII_phages_clustering.R
# 第98,141,142,153行,将原分类注释表中的Genus删掉,然后换成Genus_like_groups
Mastertable$Genus <- NULL
colnames(Mastertable)[which(names(Mastertable) == "Genus_like_groups")] <- "Genus"

####################################
# 1.5.1 Create a matrix of abundance and taxonomy tables and sample_data
# 1.5.1.1 提取丰度列然后创建矩阵
abundance_table_rarefied <- Mastertable[, grep("^raw_counts_", colnames(Mastertable))]
abundance_table_rarefied_m <- as.matrix(abundance_table_rarefied)

# 1.5.1.2 提取分类列然后创建矩阵 
target_cols <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Subfamily", "Genus", "Species")
#target_cols <- c("Genus")
valid_cols <- target_cols[target_cols %in% colnames(Mastertable)]
taxonomy_table_rarefied <- Mastertable[, valid_cols, drop = FALSE]
taxonomy_table_rarefied_m <- as.matrix(taxonomy_table_rarefied)
# 显示最终的列名
cat("taxonomy tables 实际存在的列:", paste(colnames(taxonomy_table_rarefied), collapse = ", "), "\n")

# 1.5.1.3 创建 sample_data
sample_table_rarefied_for_all <- data.frame(row.names = colnames(abundance_table_rarefied))
# group: 先去掉行名最后一个下划线及之后的部分,再去掉最后一个下划线及之前的部分
sample_table_rarefied_for_all$group <- sub(".*_([^_]+)_.*", "\\1", rownames(sample_table_rarefied_for_all))

####################################
# 1.5.2 Transform to phylseq objects
ABUNDANCE_rarefied <- otu_table(abundance_table_rarefied_m, taxa_are_rows = TRUE)
TAX_rarefied <- tax_table(taxonomy_table_rarefied_m)
samples <- sample_data(sample_table_rarefied_for_all)
phyloseq_rarefied_phages <- phyloseq(ABUNDANCE_rarefied,TAX_rarefied, samples)

####################################
# 1.5.3 Visualize data & subset only the phage fraction
#sample_names(phyloseq_rarefied_phages) # All sample names
#rank_names(phyloseq_rarefied_phages) # All taxonomies
#sample_variables(phyloseq_rarefied_phages) ## All metadata
#phyloseq_rarefied_phages
```

```{r}
#################################### 2.1 Data trimming (aggregate common taxa) ####################################
# 将数据聚合到属水平
phyloseq_rarefied_phages_visualization <- aggregate_taxa(phyloseq_rarefied_phages, 'Genus')

# 对数据进行转换和稀有分类单元过滤
phyloseq_rarefied_phages_DMM <- phyloseq_rarefied_phages_visualization %>%
  microbiome::transform(transform = "identity") %>%
  microbiome::aggregate_rare(
    level = "Genus",
    detection = 0.01 / 100,
    prevalence = 20 / 100
  )

# 检查聚合前后的变化
cat("聚合前分类单元数:", nrow(tax_table(phyloseq_rarefied_phages)), "\n")
cat("聚合后分类单元数:", nrow(tax_table(phyloseq_rarefied_phages_visualization)), "\n")

# 检查属水平的唯一值
cat("聚合后唯一属数:", length(unique(tax_table(phyloseq_rarefied_phages_visualization)[,"Genus"])), "\n")

# -----------------------------
# 新增：统计低丰度/低 prevalence 被过滤掉的行/列
# -----------------------------

# 聚合前 OTU 表
otu_before <- otu_table(phyloseq_rarefied_phages_visualization)

# 过滤后 OTU 表
otu_after  <- otu_table(phyloseq_rarefied_phages_DMM)

# 行（Genus）变化
rows_removed <- nrow(otu_before) - nrow(otu_after)
cat("被低丰度/低 prevalence 过滤掉的行数（Genus 数）:", rows_removed, "\n")

# 列（samples）变化（一般不会变化）
cols_removed <- ncol(otu_before) - ncol(otu_after)
cat("被过滤掉的列数（样本数）:", cols_removed, "\n")
```

```{r}
#################################### 2.2 DMM algorithm ####################################
# 从phyloseq对象phyloseq_rarefied_phages_DMM中提取丰度矩阵
# dat现在包含了每个分类单元（taxa）在每个样本中的丰度数据
dat <- abundances(phyloseq_rarefied_phages_DMM)
count <- as.data.frame(t(dat))

## Remove 'Other' meaning all the unique viruses appearing only in less than 20% of samples (0.01 % abundance)
# 计算每行（样本）的总丰度并添加为新列total
count$total <- rowSums(count)
# 移除总丰度为0的行（即没有任何病毒序列的样本)
count <- count[!count$total == 0,]
count$total <- NULL
count$Other <- NULL

## Remove samples composed of zero shared viruses
count$total <- rowSums(count)
count <- count[!count$total == 0,]
sort(rowSums(count))
count$total <- NULL

## How many samples and reads are left for determining DMM clusters?
#nrow(count)
#sum(count)

## Median genera per sample using community typing
#count_1 <- count
# 将所有非零值转换为1，创建存在/不存在矩阵
#count_1[count_1 > 0] <- 1
#count_1 <- as.data.frame(count_1)
#count_1$prevalence <- rowSums(count_1)
#ncol(count_1) # 18 genera shared over 20% of samples with 0.01% abundance
#median(count_1$prevalence) # Median of 6 samples per samples
#mean(count_1$prevalence)
#nrow(count_1) # 363 samples can be used for clusters
#nrow(count_1)/377)*100 # or > 96% can be used.

# Determine DMM clusters and evaluate BIC score
count <- as.matrix(count)
fit <- lapply(1:3, dmn, count = count, verbose=TRUE)

# 选择哪个准则？
# BIC: 通常在样本量较大时更受青睐，对模型复杂度惩罚更重
# AIC: 在预测准确性方面表现较好，倾向于选择较复杂的模型
# Laplace: 适用于贝叶斯模型比较
# 在实际应用中，通常会选择多个准则中结果一致的模型，或者根据具体研究目的选择最合适的准则。

lplc <- sapply(fit, laplace) # AIC / BIC / Laplace
aic  <- sapply(fit, AIC) # AIC / BIC / Laplace
bic  <- sapply(fit, BIC) # AIC / BIC / Laplace
cat("lplc:", lplc, "\n")
cat("aic:", aic, "\n")
cat("bic:", bic, "\n")
plot(bic, type="b", xlab="Number of Dirichlet Components", ylab="Model Fit")
lines(lplc, type="b", lty = 2)
lines(aic, type="b", lty = 3)
```

```{r}
#################################### 2.2 Community typing ####################################
# 选择最佳模型
k <- which.min(unlist(bic))
k <- 2
best <- fit[[k]]

# 显示最佳模型中各聚类的混合权重
#mixturewt(best)
# 显示最佳模型中每个样本的聚类分配
cat("最佳模型中每个样本的聚类分配:\n")
best@group
# 显示最佳模型的混合参数信息
#best@mixture

# 此外，评估样本属于两个DMM组中某一个的概率。如果不太感兴趣可以省略这部分
# best@group 包含了每个样本在各个聚类中的概率信息
#group_probability <- as.data.frame(best@group)
#mean(group_probability$V1)  # 计算第1个聚类的平均概率
#mean(group_probability$V2)  # 计算第2个聚类的平均概率
#median(group_probability$V1)  # 计算第1个聚类的中位数概率
#median(group_probability$V2)  # 计算第2个聚类的中位数概率

# 根据概率选择每个样本所属的主要聚类（概率最高的那个）
# mixture(best) 获取混合模型中每个样本在各聚类的概率
cluster <- apply(mixture(best), 1, which.max)  # 对每行应用which.max函数，找出最大概率对应的聚类
cluster <- as.data.frame(cluster)  # 转换为数据框格式
cluster$cluster <- as.character(cluster$cluster)  # 将聚类列转换为字符型
#cat("每个聚类的样本数:\n")
#table(cluster$cluster) # 统计每个聚类的样本数
cluster_phylo <- sample_data(cluster)  # 将聚类结果转换为phyloseq样本数据格式，用于后续分析

# 将聚类结果添加到phyloseq对象中
phyloseq_rarefied_phages_visualization_1 <- merge_phyloseq(phyloseq_rarefied_phages_visualization, cluster_phylo)

# 保存对象到文件
dir.create("03_community_types", showWarnings = FALSE)
saveRDS(
  phyloseq_rarefied_phages_visualization_1,
  file = "03_community_types/phyloseq_rarefied_phages_visualization_1.rds"
)
cat("合并肠型分析结果的 phyloseq 对象已保存到 03_community_types/phyloseq_rarefied_phages_visualization_1.rds 中 \n")
```

```{}
# At last find drivers or contribution of each taxonomic group to each component
# 最后找出每个分类群对每个成分的驱动因素或贡献
for (k in seq(ncol(fitted(best)))) {  # 遍历每个聚类（成分）
  d <- melt(fitted(best))  # 将拟合结果转换为长格式数据
  colnames(d) <- c("OTU", "cluster", "value")  # 重命名列：OTU、聚类、值
  d <- subset(d, cluster == k) %>%  # 筛选当前聚类k的数据
    # Arrange OTUs by assignment strength
    # 按照分配强度对OTU进行排序
    arrange(value) %>%
    mutate(OTU = factor(OTU, levels = unique(OTU))) %>%  # 将OTU转换为因子并设置水平
    # Only show the most important drivers
    # 仅显示最重要的驱动因素（取绝对值最高的1%）
    filter(abs(value) > quantile(abs(value), 0.99)) #0.8 default     
  
  p <- ggplot(d, aes(x = OTU, y = value)) +  # 创建柱状图
    geom_bar(stat = "identity") +  # 使用原始值绘制柱状图
    coord_flip() +  # 翻转坐标轴（水平显示）
    labs(title = paste("Top drivers: community type", k))  # 设置图表标题
  print(p)  # 打印图表
}

# 分析病毒属对不同DMM聚类的贡献程度
contribution_viruses <- as.data.frame(fitted(best))
contribution_viruses$cluster1 <- contribution_viruses$V1
contribution_viruses$cluster2 <- contribution_viruses$V2
contribution_viruses$V1 <- NULL
contribution_viruses$V2 <- NULL
```

```{}
###################################
# 2.3 Add clusters to phyloseq object
# 将聚类结果添加到phyloseq对象中，并计算样本中的属水平多样性信息
###################################
# 合并聚类信息和噬菌体phyloseq对象
phyloseq_rarefied_phages_visualization_1 <- merge_phyloseq(phyloseq_rarefied_phages_visualization, cluster_phylo)
###################################
# 2.4 Determine number of genera in samples
###################################
# How was a genus defined again?
# ----> determined as in https://github.com/snayfach/MGV/tree/master/aai_cluster
# ----> genera are binned and defined on similarity of more or equal than 20% amino acid identity (%AAI) between genomes.
# ----> and genomes with either 8 shared genes or at least 20% of shared genes (relative to both genomes)
# ----> So basically based on generic %AAI cutoff for phages

# 提取合并后phyloseq对象中的丰度数据,转换为存在/不存在矩阵
genera <- as.data.frame(t(abundances(phyloseq_rarefied_phages_visualization_1)))
genera[genera > 0] <- 1
genera$prevalence <- rowSums(genera) # 计算每个样本中的属数（流行度）

# generic infiormation
# A) Min and maximum genera in samples 计算所有样本中属数的范围（最小值和最大值）
range(genera$prevalence) # 3 min; 74 maximum

# B) genera in total 计算总的属数量
nrow(tax_table(phyloseq_rarefied_phages_visualization_1)) # 874 genera

# C) Median genera per sample 每个样本中属数的中位数（26）和平均数（27.23）
median(genera$prevalence) # 26 median
mean(genera$prevalence) # 27.23 average
###################################
```
